src/web/page/pages.rs

```diff
@@ -0,0 +1,72 @@
+use super::TEMPLATE_DATA;
+use crate::docbuilder::Limits;
+use iron::{
+    headers::ContentType,
+    mime::{Mime, SubLevel, TopLevel},
+    response::Response,
+    status::Status,
+    IronResult,
+};
+use serde::Serialize;
+use tera::Context;
+
+/// The central trait that rendering pages revolves around, it handles selecting and rendering the template
+pub trait WebPage: Serialize + Sized {
+    /// Turn the current instance into a `Response`, ready to be served
+    // TODO: We could cache similar pages using the `&Context`
+    fn into_response(self) -> IronResult<Response> {
+        let ctx = Context::from_serialize(&self).unwrap();
+        let rendered = TEMPLATE_DATA
+            .templates
+            .load()
+            .render(Self::template(), &ctx)
+            .unwrap();
+
+        let mut response = Response::with((self.get_status(), rendered));
+        response.headers.set(Self::content_type());
+
+        Ok(response)
+    }
+
+    /// Get the name of the template to be rendered
+    fn template() -> &'static str;
+
+    /// Gets the status of the request, defaults to `Ok`
+    fn get_status(&self) -> Status {
+        Status::Ok
+    }
+
+    /// The content type that the template should be served with, defaults to html
+    fn content_type() -> ContentType {
+        ContentType::html()
+    }
+}
+
+/// The sitemap, corresponds with `templates/sitemap.xml`
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub(crate) struct SitemapXml {
+    /// The release's names and RFC 3339 timestamp to be displayed on the sitemap
+    pub releases: Vec<(String, String)>,
+}
+
+impl WebPage for SitemapXml {
+    fn template() -> &'static str {
+        "docsrs/sitemap.xml"
+    }
+
+    fn content_type() -> ContentType {
+        ContentType(Mime(TopLevel::Application, SubLevel::Xml, vec![]))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub(crate) struct About {
+    pub rustc_version: Option<String>,
+    pub limits: Limits,
+}
+
+impl WebPage for About {
```

Comment:
Thinking about these impls, could a small macro be more clean?

```rust
impl_webpage!(About, "docsrs/about.html");
impl_webpage!(NotFound, "docsrs/about.html", status = NotFound);
impl_webpage!(SitemapXml, "docsrs/sitemap.xml", content_type = ContentType(Mime(TopLevel::Application, SubLevel::Xml, vec![])));
```

I'm not fully sold on using the macro myself, but it might be an interesting idea.

Reply:
Ohh, I really like that

Reply:
It has been done, 10/10 idea